src/lsm/manifest_level.zig

```diff
@@ -38,6 +38,23 @@ pub fn ManifestLevelType(
 
         pub const Tables = SegmentedArray(TableInfo, NodePool, table_count_max, .{});
 
+        pub const LeastOverlapTable = struct {
+            table: *TableInfo,
+            range: OverlapRange,
+        };
+        pub const OverlapRange = struct {
+            /// The minimum key across both levels.
+            key_min: Key,
+            /// The maximum key across both levels.
+            key_max: Key,
+            // References to tables in level B that intersect with the chosen table in level A.
+            tables: std.BoundedArray(*TableInfo, constants.lsm_growth_factor),
+        };
+        pub const TableInfoPtr = union(enum) { copy: *TableInfo, from_level: *TableInfo };
+        pub const TableInfoPtrConst = union(enum) {
+            copy: *const TableInfo,
+            from_level: *const TableInfo,
+        };
```

Comment:
If we were to use _internal_ & _external_ to categorize *TableInfo, then all invocations to insert_table & remove_table use _external *const TableInfo_. With my changes, compaction would now use an _internal *TableInfo_ to update snapshot_max.

Since the only place _internal *TableInfo_  is being used is _update_table_, I didn't understand how we can leverage the generation number to verify that an invalid internal reference isn't ever used! Could you please expand on it a bit? I would also be happy to jump on a quick call to understand things more clearly. :D

Reply:
This code isn't 100% in my head so maybe this wouldn't work.
But what I was imagining is:

- Add a `generation: usize` field to `ManifestLevel` (initially `0`). 
- When we are querying the compaction range, include the current value of `generation` in the result.
- Whenever `ManifestLevel.insert_table`/`remove_table` is called, increment the `ManifestLevel`'s `generation`.
- In `ManifestLevel.set_snapshot_max`, assert that the parameter's `generation` matches the `ManifestLevel`'s own generation. If they don't match, it means that there was a insert/remove in the mean time, so the internal pointer is not valid.

`contains()` checks that the pointer is indeed internal, but it doesn't check that the underlying table being pointed to is the same that it was when the compaction first started.

Reply:
Ah yes, a monotonically increasing generation number would indeed be more robust. Thanks a lot for expanding on the idea, dj!